File: Android/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/viewcheck/ViewCheckDokitView.java
package com.didichuxing.doraemonkit.kit.viewcheck;

import android.app.Activity;
import android.content.Context;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.LinearLayout;

import com.blankj.utilcode.util.ActivityUtils;
import com.didichuxing.doraemonkit.DoraemonKit;
import com.didichuxing.doraemonkit.R;
import com.didichuxing.doraemonkit.ui.base.AbsDokitView;
import com.didichuxing.doraemonkit.ui.base.DokitViewLayoutParams;
import com.didichuxing.doraemonkit.ui.layoutborder.ViewBorderFrameLayout;
import com.didichuxing.doraemonkit.util.LifecycleListenerUtil;
import com.didichuxing.doraemonkit.util.LogHelper;
import com.didichuxing.doraemonkit.util.UIUtils;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by jintai on 2019/09/26.
 */

public class ViewCheckDokitView extends AbsDokitView {
    private static final String TAG = "ViewCheckFloatPage";


    private List<OnViewSelectListener> mViewSelectListeners = new ArrayList<>();

    private Activity mResumedActivity;

    /**
     * 监听每个悬浮窗的activity生命周期
     */
    private LifecycleListenerUtil.LifecycleListener mLifecycleListener = new LifecycleListenerUtil.LifecycleListener() {
        @Override
        public void onActivityResumed(Activity activity) {
            mResumedActivity = activity;
            View selectView;
            if (isNormalMode()) {
                selectView = findSelectView(getNormalLayoutParams().leftMargin + getRootView().getWidth() / 2, getNormalLayoutParams().topMargin + getRootView().getHeight() / 2);
            } else {
                selectView = findSelectView(getSystemLayoutParams().x + getRootView().getWidth() / 2, getSystemLayoutParams().y + getRootView().getHeight() / 2);

            }
            onViewSelected(selectView);
        }

        @Override
        public void onActivityPaused(Activity activity) {

        }

        @Override
        public void onFragmentAttached(Fragment f) {

        }

        @Override
        public void onFragmentDetached(Fragment f) {

        }
    };

    @Override
    public void onCreate(Context context) {
        mResumedActivity = ActivityUtils.getTopActivity();
        LifecycleListenerUtil.registerListener(mLifecycleListener);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        LifecycleListenerUtil.unRegisterListener(mLifecycleListener);
    }

    @Override
    public View onCreateView(Context context, FrameLayout view) {
        return LayoutInflater.from(context).inflate(R.layout.dk_float_view_check, null);
    }


    @Override
    public void initDokitViewLayoutParams(DokitViewLayoutParams params) {
        params.x = UIUtils.getWidthPixels(getContext()) / 2;
        params.y = UIUtils.getHeightPixels(getContext()) / 2;
        params.height = DokitViewLayoutParams.WRAP_CONTENT;
        params.width = DokitViewLayoutParams.WRAP_CONTENT;
    }

    @Override
    public void onViewCreated(FrameLayout view) {

    }

    private View findSelectView(int x, int y) {
        if (mResumedActivity == null) {
            return null;
        }
        if (mResumedActivity.getWindow() == null) {
            return null;
        }
        if (isNormalMode()) {
            //LogHelper.d(TAG, "x: " + x + ", y: " + y);
            return traverseViews(UIUtils.getDokitAppContentView(mResumedActivity), x, y);
        } else {
            return traverseViews(mResumedActivity.getWindow().getDecorView(), x, y);
        }

    }

    private View traverseViews(View view, int x, int y) {
        if (view == null) {
            return null;
        }
        int[] location = new int[2];
        view.getLocationInWindow(location);
        int left = location[0];
        int top = location[1];
        int right = left + view.getWidth();
        int bottom = top + view.getHeight();
        if (view instanceof ViewGroup) {
            int childCount = ((ViewGroup) view).getChildCount();
            if (childCount != 0) {
                for (int index = childCount - 1; index >= 0; index--) {
                    View v = traverseViews(((ViewGroup) view).getChildAt(index), x, y);
                    if (v != null) {
                        return v;
                    }
                }
            }
            if (left < x && x < right && top < y && y < bottom) {
                return view;
            } else {
                return null;
            }
        } else {
            if (view != null) {
//                LogHelper.i(TAG, "class: " + view.getClass() + ", left: " + left
//                        + ", right: " + right + ", top: " + top + ", bottom: " + bottom);
            }

            if (left < x && x < right && top < y && y < bottom) {
                return view;
            } else {
                return null;
            }
        }
    }

    public void setViewSelectListener(OnViewSelectListener viewSelectListener) {
        mViewSelectListeners.add(viewSelectListener);
        //每次新增监听的时候开始查找
        View selectView = null;
        if (isNormalMode()) {
            selectView = findSelectView(getNormalLayoutParams().leftMargin + getRootView().getWidth() / 2, getNormalLayoutParams().topMargin + getRootView().getHeight() / 2);
        } else {
            selectView = findSelectView(getSystemLayoutParams().x + getRootView().getWidth() / 2, getSystemLayoutParams().y + getRootView().getHeight() / 2);

        }
        onViewSelected(selectView);
    }

    public void removeViewSelectListener(OnViewSelectListener viewSelectListener) {
        mViewSelectListeners.remove(viewSelectListener);
    }


    @Override
    public void onUp(int x, int y) {
        super.onUp(x, y);
        View selectView = null;
        if (isNormalMode()) {
            selectView = findSelectView(getNormalLayoutParams().leftMargin + getRootView().getWidth() / 2, getNormalLayoutParams().topMargin + getRootView().getHeight() / 2);
        } else {
            selectView = findSelectView(getSystemLayoutParams().x + getRootView().getWidth() / 2, getSystemLayoutParams().y + getRootView().getHeight() / 2);

        }
        onViewSelected(selectView);
    }

    @Override
    public void onDown(int x, int y) {

    }

    private void onViewSelected(View view) {
        for (OnViewSelectListener listener : mViewSelectListeners) {
            listener.onViewSelected(view);
        }
    }

    public interface OnViewSelectListener {
        void onViewSelected(View view);
    }

}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/viewcheck/ViewCheckDokitView.kt
package com.didichuxing.doraemonkit.kit.viewcheck

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.HandlerThread
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.fragment.app.Fragment
import com.blankj.utilcode.util.ActivityUtils
import com.didichuxing.doraemonkit.R
import com.didichuxing.doraemonkit.kit.core.AbsDokitView
import com.didichuxing.doraemonkit.kit.core.DokitViewLayoutParams
import com.didichuxing.doraemonkit.util.LifecycleListenerUtil
import com.didichuxing.doraemonkit.util.UIUtils
import java.util.*

/**
 * @author: xuchun
 * @time: 2020/6/4 - 14:47
 * @desc: 控件检查View
 */
class ViewCheckDokitView : AbsDokitView(), LifecycleListenerUtil.LifecycleListener {
    private lateinit var mFindCheckViewRunnable: FindCheckViewRunnable
    private lateinit var mTraverHandlerThread: HandlerThread
    private lateinit var mTraverHandler: Handler
    private val mViewSelectListeners: MutableList<OnViewSelectListener> = ArrayList()
    private var mResumedActivity: Activity? = null
    override fun onCreate(context: Context?) {
        mTraverHandlerThread = HandlerThread(TAG)
        mTraverHandlerThread.start()
        mTraverHandler = Handler(mTraverHandlerThread.looper)
        mFindCheckViewRunnable = FindCheckViewRunnable()
        mResumedActivity = ActivityUtils.getTopActivity()
        LifecycleListenerUtil.registerListener(this)
    }

    override fun onDestroy() {
        super.onDestroy()
        mTraverHandler.removeCallbacks(mFindCheckViewRunnable)
        mTraverHandlerThread.quit()
        LifecycleListenerUtil.unRegisterListener(this)
    }

    override fun onCreateView(context: Context?, rootView: FrameLayout?): View {
        return LayoutInflater.from(context).inflate(R.layout.dk_float_view_check, null)
    }

    override fun initDokitViewLayoutParams(params: DokitViewLayoutParams?) {
        params?.let {
            it.x = UIUtils.widthPixels / 2
            it.y = UIUtils.heightPixels / 2
            it.height = DokitViewLayoutParams.WRAP_CONTENT
            it.width = DokitViewLayoutParams.WRAP_CONTENT
        }

    }

    override fun onUp(x: Int, y: Int) {
        super.onUp(x, y)
        preformFindCheckView()
    }

    override fun onActivityResumed(activity: Activity?) {
        mResumedActivity = activity
        preformFindCheckView()
    }

    override fun onViewCreated(rootView: FrameLayout?) {}
    override fun onActivityPaused(activity: Activity?) {}
    override fun onFragmentAttached(f: Fragment?) {}
    override fun onFragmentDetached(f: Fragment?) {}
    override fun onDown(x: Int, y: Int) {}
    fun setViewSelectListener(viewSelectListener: OnViewSelectListener) {
        mViewSelectListeners.add(viewSelectListener)
        preformFindCheckView()
    }

    fun removeViewSelectListener(viewSelectListener: OnViewSelectListener) {
        mViewSelectListeners.remove(viewSelectListener)
    }

    fun preformPreCheckView() {
        mFindCheckViewRunnable.mIndex--
        if (mFindCheckViewRunnable.mIndex < 0) {
            mFindCheckViewRunnable.mIndex += mFindCheckViewRunnable.mCheckViewList!!.size
        }
        mFindCheckViewRunnable.dispatchOnViewSelected()
    }

    fun preformNextCheckView() {
        mFindCheckViewRunnable.mIndex++
        if (mFindCheckViewRunnable.mIndex >= mFindCheckViewRunnable.mCheckViewList!!.size) {
            mFindCheckViewRunnable.mIndex -= mFindCheckViewRunnable.mCheckViewList!!.size
        }
        mFindCheckViewRunnable.dispatchOnViewSelected()
    }

    private fun preformFindCheckView() {
        val x: Int
        val y: Int
        if (isNormalMode) {
            x = normalLayoutParams.leftMargin + rootView.width / 2
            y = normalLayoutParams.topMargin + rootView.height / 2
        } else {
            x = systemLayoutParams.x + rootView.width / 2
            y = systemLayoutParams.y + rootView.height / 2
        }
        mTraverHandler.removeCallbacks(mFindCheckViewRunnable)
        mFindCheckViewRunnable.mX = x
        mFindCheckViewRunnable.mY = y
        mTraverHandler.post(mFindCheckViewRunnable)
    }

    private fun traverseViews(viewList: MutableList<View>, view: View?, x: Int, y: Int) {
        if (view == null) {
            return
        }
        val location = IntArray(2)
        view.getLocationInWindow(location)
        val left = location[0]
        val top = location[1]
        val right = left + view.width
        val bottom = top + view.height

        // 深度优先遍历
        if (view is ViewGroup) {
            val childCount = view.childCount
            if (childCount != 0) {
                for (index in childCount - 1 downTo 0) {
                    traverseViews(viewList, view.getChildAt(index), x, y)
                }
            }
        }
        if (x in (left + 1) until right && top < y && y < bottom) {
            viewList.add(view)
        }
    }

    private fun onViewSelected(current: View?, checkViewList: List<View>?) {
        for (listener in mViewSelectListeners) {
            listener.onViewSelected(current, checkViewList!!)
        }
    }

    interface OnViewSelectListener {
        fun onViewSelected(current: View?, checkViewList: List<View>)
    }

    internal inner class FindCheckViewRunnable : Runnable {
        var mX = 0
        var mY = 0
        var mIndex = 0
        var mCheckViewList: List<View>? = null
        override fun run() {
            val viewList: MutableList<View> = ArrayList(20)
            if (mResumedActivity != null && mResumedActivity!!.window != null) {
                if (isNormalMode) {
                    //LogHelper.d(TAG, "x: " + mX + ", y: " + mY);
                    traverseViews(viewList, UIUtils.getDokitAppContentView(mResumedActivity), mX, mY)
                } else {
                    traverseViews(viewList, mResumedActivity!!.window.decorView, mX, mY)
                }
            }
            mIndex = 0
            mCheckViewList = viewList
            dispatchOnViewSelected()
        }

        fun dispatchOnViewSelected() {
            post(Runnable { onViewSelected(currentCheckView, mCheckViewList) })
        }

        private val currentCheckView: View?
            get() {
                val size = mCheckViewList!!.size
                return if (size == 0) {
                    null
                } else mCheckViewList!![mIndex]
            }
    }

    companion object {
        private const val TAG = "ViewCheckFloatPage"
    }
}